zamba/models/depth_estimation/depth_manager.py

```diff
@@ -0,0 +1,242 @@
+
+                # if anything is detected in the frame, save out relevant frames
+                if len(detections) > 0:
+                    logger.debug(f"{len(detections)} detection(s) found at second {frame_idx}.")
+
+                    # get frame indices around frame with detection
+                    min_frame = frame_idx - self.window_size
+                    max_frame = frame_idx + self.window_size
+
+                    # add relevant resized frames to dict if not already added
+                    # if index is before start or after end of video, use an array of zeros
+                    for i in range(min_frame, max_frame + 1):
+                        if f"frame_{i}" not in cached_frames[video_filepath].keys():
+                            try:
+                                selected_frame = np.array(
+                                    # PIL expects size to be (w, h)
+                                    Image.fromarray(arr[i]).resize((self.width, self.height))
+                                )
+                            except:
+                                selected_frame = np.zeros((self.height, self.width, self.channels))
+
+                            cached_frames[video_filepath][f"frame_{i}"] = selected_frame
+
+                    # iterate over detections in frame to create universal detection ID
+                    for i, (detection, score) in enumerate(zip(detections, scores)):
+                        universal_det_id = f"{i}_{frame_idx}_{video_filepath}"
+
+                        # save out bounding box and score info in case we want to mask out portions
+                        detection_dict[universal_det_id] = dict(
+                            bbox=detection,
+                            score=score,
+                            frame=frame_idx,
+                            video=video_filepath,
+                        )
+
+            del arr
+
+        self.detection_dict = detection_dict
+        self.detection_indices = list(detection_dict.keys())
+        self.cached_frames = cached_frames
+
+    def __len__(self):
+        return len(self.detection_indices)
+
+    def __getitem__(self, index):
+        """Given the index of the target image, returns a tuple of the stacked image array, the image
+        filename stem, and the time into the video for the target image.
+        """
+
+        # get detection info
+        detection_idx = self.detection_indices[index]
+        det_metadata = self.detection_dict[detection_idx]
+        det_frame = det_metadata["frame"]
+        det_video = det_metadata["video"]
+
+        # set up input array of frames within window of detection
+        # frames are stacked channel-wise
+        input = np.zeros((self.height, self.width, self.channels * self.num_frames))
+        n = 0
+        for frame_idx in range(det_frame - self.window_size, det_frame + self.window_size + 1):
+            input[:, :, n : n + self.channels] = self.cached_frames[det_video][
```

Comment:
worth looking into ways to be more efficient in the memory management here (e.g. no allocating ahead and doing the stack)